README.md
# PLL: Programming in Logic in Lisp

PLL is a set of implementations of Prolog in Scheme. It is not
supposed to be efficient, but rather as a simple way to teach the
fundamentals of Logic Programming and the internal working of a
Prolog interpreter (conceptually only -- a real implementation would
be radically different). This interpreter is based on a simple
implementation of the AMB operator.

PLL is very small (less than 500 lines of code, including different
versions of the interpreter).

Each variant is built on top of the other.

- Pure prolog: no variables, no assertions, only plain Prolog and
  SLD-resolution.

- Prolog w/built-ins: with an extensible set of built-in predicates.
  Only the built-ins within a list are allowed.

- Prolog w/Scheme functions: call any Scheme function from Prolog.

- Prolog w/local vars: this version has support for "IS" and local
  Prolog variables.

- Prolog w/meta-predicates: this version has support for "assert"
  and "retract".

- Prolog w/cut: this version supports cuts.

One last implementation is missing, that would allow for writing
and reading the database (so it would be possible to store and
retrieve programs). This is very simple to implement and planned
for the near future.

This interpreter is implemented in three files:

- `prolog-core.scm`: this is the main program, where all the above variants
  of Prolog are implemented. Each variant differs from the previous
  only slightly, and the difference is made clear in code comments.

- `amb.scm`: handles non-determinism. This is a simple and usual
  implementation of the AMB operator.

- `unify.scm`: contains code for unification.

There are also two files used to load the interpreter:

- `pll.scm`: contains an R7RS library declaration (for Chibi users:
  `pll.sld` is a symlink to `pll.scm`, and works as expected)

- `pll-standalone.scm`: contains code to load the interpreter,
   without creating any libraries or modules. Will override
   identifiers!

## Supported Scheme implementations


I have made an effort to get the interpreter working in as many Scheme
implementations as I could. Below is a list of supported and not
supported Schemes. "Supported" means I ran the example code pieces
(in prolog-examples.scm) and they worked as sxpected. I did no
further tests.

There are two ways to use PLL:

```
(load "pll-standalone.scm")    ;; this works on all supported schemes.

(import (pll))                 ;; only on the systems marked as having 
                               ;; support for R7RS modules
```

```
Scheme      version	R7RS module	Note
-----------------------------------------------
Chicken     4.*		Y
Gauche      0.9.4	Y
Guile       2.*		-
Chibi       0.7.2	Y
Husk        3.19.1	-
SISC        1.16.6	-
Saggitarius 0.7.1	Y
Gambit      4.7		-		Needs -:s flag for syntax rules, and SRFI-1
MIT Scheme  9.1.1	-
STklos      1.10	-
Scheme48       	 	-		Needs ,open srfi-1
```

Currently not supported:

```
Bigloo     (prolog+cut eats up 100% CPU; needs investigation)
SigScheme  (syntax-rules is disabled in sigscheme for now)
TinyScheme (syntax-rules not implemented)
Scheme9    (very strange behavior -- needs investigation)
Kawa       (AMB doesn't work because of some different call/cc behavior)
Foment     (the Hanoi towers example eats up 100% CPU and never finishes)
SCM        (support for neither cond-expand nor syntax-rules)
SIOD       (syntax-rules not implemented)
```

## Documentation

A very short manual is included (manual.md, manual.txt, manual.pdf).



changelog
1.6.3
* Partial support for saving and restoring built-in procedures
1.6.2
* Fixed one small packaging bug
1.6
* Scheme48 is supported
* Fixed Chicken packaging problems
* Documentation cleanup
1.5
* Packaging: now available as R7RS module for some implementations;
  included instructions on how to load, with and without R7RS
* Manual in Markdown; translated into HTML, txt and PDF
* Now available as Chicken Egg
1.4
* Minor changes to the manual and README
* Fixing licence headers
1.3
* Compatibility with STklos
1.2
* Fixed compatibility with MIT, SISC, and Gambit, while also
  cleaning up the code that does implementation-specific stuff.
1.1
* Initial release


amb.scm
;; This software is Copyright (c) Jeronimo C. Pellegrini, 2010-2016.
;;
;;  This program is free software; you can redistribute it and/or
;;  modify it under the terms of the GNU General Public License as
;;  published by the Free Software Foundation; either version 3 of the
;;  License, or (at your option) any later version.
;; 
;;  This program is distributed in the hope that it will be useful, but
;;  WITHOUT ANY WARRANTY; without even the implied warranty of
;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;;  General Public License for more details.
;; 
;;  You should have received a copy of the GNU General Public License
;;  along with this program; if not, write to:
;;    The Free Software Foundation, Inc.,
;;    51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

;; amb.scm -- implements the AMB operator
;; 
;; SYNOPSIS:

;;; amb
;;;
;; (amb-reset)
;; :: resets the amb tree: this will make amb forget everything.
;;
;; (amb x1 x2 ...)
;; :: add "x1 x2 ..." to the set of answers
;;
;; (amb)
;; :: selects one of the answers
;;
;; (require p)
;; :: selects one of the answers x such that (p x) is #true
;;
;; (amb-list lst)
;; :: same as (amb x1 x2 ...), where lst is (list x1 x2 ...)
;;
;; (amb-all v)
;; :: shows all values that may still be returned by amb in the future
;;
;; if amb has no more solutions to show, it will signal an error


;;; amb+
;;;
;; (amb+-reset)
;; :: resets the amb tree: this will make amb+ forget everything.
;;
;; (amb+ x1 x2 ...)
;; :: add "x1 x2 ..." to the set of answers
;;
;; (amb+)
;; :: selects one of the answers
;;
;; (require+ p)
;; :: selects one of the answers x such that (p x) is #true
;;
;; (amb+-delete-one)
;; :: deletes one element from the amb+ set.
;;
;; (amb+-add k)
;; :: adds k to the amb set
;;
;; (amb+-list lst)
;; :: same as (amb+ x1 x2 ...), where lst is (list x1 x2 ...)
;;
;; (amb+-fail)
;; :: synonym to (amb+)
;;
;; if amb+ has no more solutions to show, it will return #f



(begin

;; called when (amb) is called and there is no more answers available.
(define fail
  (lambda ()
    (error "Amb tree exhausted"))) 

(define amb-saved-fail fail)

;; resets the amb tree: this will make amb forget everything.
(define amb-reset
  (lambda ()
    (set! fail amb-saved-fail)))



(define-syntax amb 
  (syntax-rules () 
    ((amb) (fail))
    ((amb expression) expression)
    ((amb expression ...)
     (let ((saved-fail fail))
       (call-with-current-continuation
        (lambda (k-success)
          (call-with-current-continuation
           (lambda (k-failure)
             (set! fail (lambda ()
                          (set! fail saved-fail)
                          (k-failure 'boo)))
             (k-success expression)))
            ...
            (saved-fail)))))))


(define amb-save
  (lambda ()
    (let ((amb-saved-fail-bak amb-saved-fail)
          (amb-fail-bak fail))
      (amb-reset)
      (list amb-saved-fail-bak amb-fail-bak))))

(define amb-restore
  (lambda (lst)
     (set! fail (car lst))
     (set! amb-saved-fail (cadr lst))))

;; gets the first value from amb that satisfies p.
(define (require p)
  (if (not p) (amb)))

;; shows all values available.
(define-syntax amb-all
  (syntax-rules ()
    ((_ e)
     (let ((saved-fail fail)
           (results '()))
       (if (call-with-current-continuation
            (lambda (k)
              (set! fail (lambda () (k #f)))
              (let ((v e))
                (set! results (cons v results))
                (k #t))))
           (fail))
       (set! fail saved-fail)
       (reverse results)))))

(define now
  (lambda ()
    (call-with-current-continuation 
     (lambda (k) (k k)))))

(define amb+-set '())

(define (amb+-reset)
  (set! amb+-set '()))

(define (amb+-delete-one)
  (set! amb+-set (cdr amb+-set)))

(define (amb+-add k)
  (set! amb+-set (cons k amb+-set)))

(define (amb+-fail)
  (if (null? amb+-set)
      #f;(error "amb+: no more choices!")
      (let ((back (car amb+-set)))
        (amb+-delete-one)
        (back back))))

(define (amb+ . args)
  (let ((choices args))
    (let ((cc (now)))
      (cond ((null? choices) (amb+-fail))
            ((pair? choices)
             (let ((choice (car choices)))
               (set! choices (cdr choices))
               (amb+-add cc)
               choice))
            (else (error "amb choices must be a list"))))))

(define (amb+-list lst) (apply amb+ lst))

(define (require+ p)
  (if (not p) (amb+)))

(define amb-list
  (lambda (lst)
    (let loop ((lst lst))
      (if (null? lst)
          (amb)
          (amb (car lst) (loop (cdr lst)))))))

) ;; begin


unify.scm
;; This software is Copyright (c) Jeronimo C. Pellegrini, 2010-2016.
;;
;;  This program is free software; you can redistribute it and/or
;;  modify it under the terms of the GNU General Public License as
;;  published by the Free Software Foundation; either version 3 of the
;;  License, or (at your option) any later version.
;; 
;;  This program is distributed in the hope that it will be useful, but
;;  WITHOUT ANY WARRANTY; without even the implied warranty of
;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;;  General Public License for more details.
;; 
;;  You should have received a copy of the GNU General Public License
;;  along with this program; if not, write to:
;;    The Free Software Foundation, Inc.,
;;    51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA


;; unify.scm -- implements unification
;; 
;; SYNOPSIS:
;; 
;; (unify x y)
;; :: unifies x and y. 

;;; UNIFICATION WITH STACK
;;; ----------------------
;;;

(begin

;; a symbol is a matching symbol if it begins with a question
;; mark.
(define matching-symbol?
  (lambda (s)
    (and (symbol? s)
         (eqv? (string-ref (symbol->string s) 0)
              #\?))))


(define apply-sub
  (lambda (sub exp)
    (cond ((eqv? exp '()) '())
          ((pair? exp)
           (cons (apply-sub sub (car exp))
                 (apply-sub sub (cdr exp))))
          (else
           (let ((new-x (assoc exp sub)))
             (if new-x
                 (cdr new-x)
                 exp))))))


(define variable?
   (lambda (s)
    (and (symbol? s)
         (eqv? (string-ref (symbol->string s) 0)
              #\?))))

(define bound? assq)

(define (value v subst) (cdr (assq v subst)))

(define (bind var val sub) (cons (cons var val) sub))

(define occurs?
  (lambda (var x subst)
    (cond ((eqv? var x) #t)
          ((bound? x subst)
           (occurs? var (value x subst) subst))
          ((pair? x) (or (occurs? var (car x) subst)
                         (occurs? var (cdr x) subst)))
          (else #f))))

(define uni-var
  (lambda (var val sub)
    (cond ((eqv? var val) sub)
          ((bound? var sub)
           (unify (value var sub) val sub))
          ((and (variable? val) (bound? val sub))
           (unify var (value val sub) sub))
          ((occurs? var val sub) #f)
          (else (bind var (apply-sub sub val)
                      (apply-sub (list (cons var val)) sub))))))

(define unify
  (lambda (x y . s)
    (let ((sub (if (null? s) '() (car s))))
      (cond ((equal? x y) sub)
            ((eqv? sub #f) #f)
            ((variable? x) (uni-var x y sub))
            ((variable? y) (uni-var y x sub))
            ((not (or (pair? x) (pair? y))) #f)
            ((and (null? x) (null? y))
             sub)
            ((or (null? x) (null? y))
             #f)
            (else (unify (cdr x) (cdr y)
                         (unify (car x) (car y) sub)))))))

)


prolog-core.scm
;; This software is Copyright (c) Jeronimo C. Pellegrini, 2010-2016.
;;
;;  This program is free software; you can redistribute it and/or
;;  modify it under the terms of the GNU General Public License as
;;  published by the Free Software Foundation; either version 3 of the
;;  License, or (at your option) any later version.
;; 
;;  This program is distributed in the hope that it will be useful, but
;;  WITHOUT ANY WARRANTY; without even the implied warranty of
;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;;  General Public License for more details.
;; 
;;  You should have received a copy of the GNU General Public License
;;  along with this program; if not, write to:
;;    The Free Software Foundation, Inc.,
;;    51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

;; prolog.scm -- core of the the Prolog interpreter

(begin 
(cond-expand (mit         (define (interaction-environment) user-initial-environment))
             (sisc        (require-extension (srfi 1)))
             (else))

;; rules are lists, so we define better names for their head and tail.
(define rule-head   car)
(define rule-tail   cdr)

;; goals are lists, so we define better names for the next subgoal to be
;; selected and the rest of the subgoal
(define select-goal car)
(define next-goals  cdr)

;; renames one variable (matching symbol), adding a suffix tag.
;;
(define rename-one
  (lambda (var tag)
    (string->symbol
     (string-append
      (symbol->string var) "_"  (number->string tag)))))

;; Recurses through a structure, renaming all variables
;; (matching symbols), adding to them a suffix tag.
;;
(define rename-vars
  (lambda (vars tag)
    (cond ((pair? vars)
           (cons (rename-vars (car vars) tag)
                 (rename-vars (cdr vars) tag)))
          ((and tag (matching-symbol? vars))
           (rename-one vars tag))
          ((eqv? vars '!)
           (list '! (cdr amb+-set)))
          (else vars))))

;; a: head of clause to be unified. Example: '(f ?x)
;;
;; prog: the program. Example:
;;     '( (a ?z) ()
;;        (f 10) (y (a b) z (c d)) )) 
;;
;; Will return SUB and CLAUSE TAIL:
;;
;; (unifying-clause '(x ?a)
;;                  '( (x 5) (p ?z 1)  ))
;; 
;; (((?a . 5)) ((p ?z 1)))
;;
;; Here ((?a . 5)) is the substitution and
;; ((p ?z 1)) is the tail of the selected clause:
;;
(define unifying-clause
  (lambda (a rule)
    (if (null? rule)
        #f
        (let ((sub (unify a (rule-head rule))))
          (if sub
              (list sub (rule-tail rule))
              #f)))))

;; (unifying-clause '(x ?a)
;;                 '( (x 5) (p ?z 1)  ))
;; => (((?a . 5)) ((p ?z 1)))

;;(unifying-clause '(x ?a)
;;                 '( (x 5) ()  ))
;; => (((?a . 5)) (()))

;; the "sub/tail" structure is a list whose first element is a
;; substitution, so we define the following:
(define sub/tail->tail cadr)
(define sub/tail->sub  car)

;; Selects only the part of a substitution which is relevant
;; for a goal.
;;
;; s: subst.
;; g: goal.
;;
;; output: another subst., only for the variables in goal.
;;
;; (select-sub '( (?x 1)
;;                (?y 2)
;;                (?z 3) )
;;             '(f ?y 9 (a ?z 10 v)))
;;
;; ==> ((?z 3) (?y 2))
;;
(define select-sub
  (lambda (sub goal)
    (define relevant?
      (lambda (s)
        (and (occurs? (car s) goal '())
             (not (matching-symbol? (cdr s))))))
    (filter relevant? sub)))


;; for example,
;;(select-sub '( (?x . 1)
;;               (?y . 2)
;;               (?z . ?w)
;;               (?w . 10)
;;               (4  . 4))
;;            '( (f ?y 9) (a ?z 10 v)) )
;; ==> ((?y . 2))

;; Checks if a Scheme object occurs anywhere inside a structure
;; -- not necessarily a list.
;; Will also return #t if the object is eqv? to the structure.
;;
(define deep-occurs?
  (lambda (obj where)
    (cond ((pair? where)
           (or (deep-occurs? obj (car where))
               (deep-occurs? obj (cdr where))))
          (else (eqv? obj where)))))



;; THIS (resolve and pure-prolog) IS ONE STEP OF SLD-RESOLUTION
;;
;; Finds a unifying clause and returns:
;; - The substitution
;; - The goal after sub is applied
;;
;; goal: one atom: (f ?a 10 (g ?z))
;;
;; program:
;;         ( (p ?x 0) ((...)
;;                     (...))
;;           (p 0 ?y) ((...)) )
;;
;;
;; goal: the goal (list of terms). Doesn't change!
;;
;; res: resolvent (LIST of what still needs to prove in order to
;;      prove the goal).
;;
;; sub: the substitution that transforms the goal into a valid instance.
;;
;; tail: after finding A <- B1, B2, ..., with A unifying with the next resolvent
;;       clause, TAIL is the tail
;;       (B1, ...).     (see theta)
;;
;; theta: the substitution that unifies the next resolvent clause with one clause
;;        of the program (see tail)
;;
(define resolve
  (lambda (res sub tag goal prog)
    ;; this is best understood after reading chapter 4 of "The Art of Prolog",
    ;; by Sterling and Shapiro.
    (if (null? res)
        ;; if res is null there is nothing else to do; select the
        ;; relevant parts of the substitution, apply it to the
        ;; resolvent and return them.
        (list (select-sub sub goal) (apply-sub sub res))
        ;; the amb+-list below will include all program rules in the
        ;; amb set.
        (let ((rule (rename-vars (amb+-list prog) tag)))
          (if rule
              (begin
                ;; the following line does some "amb-magic". The
                ;; "require+" will find a rule whose head unifies with
                ;; the current goal.
                (require+ (unify (select-goal res) (rule-head rule)))
                (let ((sub+tail
                       (unifying-clause (select-goal res) rule)))
                  (let ((tail  (sub/tail->tail sub+tail))
                        (theta (sub/tail->sub sub+tail)))
                    ;; we apply theta to both the goal and to the
                    ;; current substitution.
                    (let ((new-sub (append theta (apply-sub theta sub))))
                      (let ((new-res
                             (apply-sub theta
                                        (append tail
                                                (next-goals res)))))
                        (list new-sub new-res))))))
              #f)))))

;; resolved is a list whose car is the substitution and whose tail
;; is the resolvent.
(define resolved->sub car)
(define resolved->res cadr)

;; this is the interpreter for pure Prolog.
(define pure-prolog
  (lambda (prog goal)
    ;; begin zero-ing the AMB tree.
    (amb+-reset)
    (let loop ((res goal)
               (sub '())
               (tag 1))
      (if (null? res)
          ;; if res is null, nothing else needs to be done. Select
          ;; the relevant part of the substitution (sub) for this goal
          ;; and return it.
          (select-sub sub goal)
          ;; if not,  then call resolve to perform one step of
          ;; resolution, and loop over with the new resolvent
          ;; and new substitution.
          (let ((resolved (resolve res sub tag goal prog)))
            (if resolved
                (loop (resolved->res resolved)
                      (resolved->sub resolved)
                      (+ tag 1))
                #f))))))


;;;
;;; WITH BUILT-INS
;;;

;; each built-in in internally a function of two arguments:
;; the first is a list of arguments to the built-in function
;; the second is the current substitution.
(define write-it
  (lambda (args sub)
    (cond ((not (null? args))
           (if (matching-symbol? (car args))
               (display (assoc (car args) sub))
               (display (car args)))
           (write-it (cdr args) sub))
          (else #t))))

(define built-ins #f)

;; a built-in procedure to save the list of built-in procedures
(define save-built-in-list
  (lambda (filename)
    (let ((out (open-output-file filename)))
      (write built-ins out)
      (close-output-port out))))

;; similarly, a procedure to load the built-in list.
(define load-built-in-list
  (lambda (filename)
    (let ((in (open-input-file filename)))
      (let ((x (read in)))
        (if (not (eof-object? x))
            (set! built-ins x)))
      (close-input-port in))))


;; this is the list of default built-ins (there are only three, WRITE, SAVE-BUILT-IN-LIST and LOAD-BUILT-IN-LIST)
(set! built-ins `((write . ,write-it)
                  (save  . ,save-built-in-list)
                  (load  . ,load-built-in-list)))

;; this will interpret a built-in.
;; proc-call is of the form (built-in-name arg1 arg2 ...)
(define interpret-built-in
  (lambda (proc-call sub)
    (let ((proc (cdr (assq (car proc-call) built-ins)))
          (args (cdr proc-call)))
      (proc args sub))))

;; res is of the form "(proc arg1 arg2 ...)"
(define built-in?
  (lambda (res)
    (assq (car res) built-ins)))

(define prolog+built-ins
  (lambda (prog goal)
    (amb-reset)
    (let loop ((res goal)
               (sub '())
               (tag 1))

      (cond ((null? res)
             (list (select-sub sub goal) (apply-sub sub res)))

            ;; if we have a built-in, then call interpret-built-in,
            ;; and loop over.
            ((built-in? (car res))
             (interpret-built-in (car res) sub)
             (loop (cdr res) sub tag))

            (else (let ((resolved (resolve res sub tag goal prog)))
                    (loop (resolved->res resolved)
                          (resolved->sub resolved)
                          (+ tag 1))))))))



;;;
;;; WITH SCHEME PROCEDURES
;;;


(define atom-start car)

(define prolog+scheme
  (lambda (prog goal)
    (amb+-reset)
    (let loop ((res goal)
               (sub '())
               (tag 1))
      (cond ((null? res)
             (select-sub sub goal))

            ;; if we have a list, it's a call to a Scheme procedure,
            ;; so call it, get the return value, require it to
            ;; be different feorm #f and loop over. If the returned
            ;; value is false, then return #f (so Prolog will fail).
            ((list? (atom-start (car res)))
             (let ((truth (eval (apply-sub sub
                                           (atom-start (car res))) 
                                (interaction-environment))))
               (require+ truth)
               (if truth
                   (loop (cdr res) sub tag)
                   #f)))

            (else (let ((resolved (resolve res sub tag goal prog)))
                    (if resolved
                        (loop (resolved->res resolved)
                              (resolved->sub resolved)
                              (+ tag 1))
                        #f)))))))




;;;
;;; WITH LOCAL VARIABLES AND "IS"
;;;

;; the rvalue of an IS is its second element (the cadr).
(define is->rvalue cadr)

;; do-is is the procedure that makes IS and local variables
;; work.
;;
;; res is of the form "(is ?var expression)"
;; do-is will try to unify ?var, which is
;; (atom-start (cdr res)), with expression,
;; which needs to be evaluated, using eval in
;; interaction-environment. We require the unification
;; to be successful, otherwise we fail (that is what
;; the "(require+ s)" does).
(define do-is
  (lambda (res sub)
    (let ((s (unify
              (atom-start (cdr res))
              (eval (apply-sub sub
                               (is->rvalue (cdr res)))
                    (interaction-environment)))))
      (require+ s)
      s)))

(define prolog+local
  (lambda (prog goal)
    (amb+-reset)
    (let loop ((res goal)
               (sub '())
               (tag 1))
      (cond ((null? res)
             (select-sub sub goal))

            ;; if we have a "IS", call 'do-is' on the first
            ;; element of res, get the substitution returned,
            ;; apply it to the rest of the goal and to the
            ;; current substitution, and loop over.
            ((eqv? (atom-start (car res)) 'is)
             (let ((s (do-is (car res) sub)))
               (loop (apply-sub s (cdr res))
                     (append s (apply-sub s sub)) tag)))

            ((list? (atom-start (car res)))
             (let ((truth (eval (apply-sub sub
                                           (atom-start (car res)))
                                (interaction-environment) )))
               (require+ truth)
               (if truth
                   (loop (cdr res) sub tag)
                   #f)))

            (else
             (let ((resolved (resolve res sub tag goal prog)))
               (if resolved
                   (loop (resolved->res resolved)
                         (resolved->sub resolved)
                         (+ tag 1))
                   #f)))))))


;;;
;;; WITH META-PREDICATES (assert, retract)
;;;

(define prolog+meta
  (lambda (prog goal)

    ;; simple appending procedure. We add the new facts to 
    ;; the global variable "prog".
    (define asserta
      (lambda (res)
        (let ((fact (cdar res)))
          (set! prog (cons fact prog)))))

    ;; simple retract procedure: just remove the fact from
    ;; "prog".
    (define retract
      (lambda (res)
        (let ((fact (cadar res)))
          (set! prog (remove (lambda (x) (equal? x fact))
                             prog)))))
    
    (amb+-reset)
    (let loop ((res goal)
               (sub '())
               (tag 1))
      (cond ((null? res)
             (select-sub sub goal))

            ;; asserta adds the fact to the database:
            ((eqv? (atom-start (car res)) 'asserta)
             (asserta res)
             (loop (cdr res) sub tag))

            ;; retract removes the fact:
            ((eqv? (atom-start (car res)) 'retract)
             (retract res)
             (loop (cdr res) sub tag))
            
            ((eqv? (atom-start (car res)) 'is)
             (let ((s (do-is (car res) sub)))
               (loop (apply-sub s (cdr res))
                     (append s (apply-sub s sub)) tag)))
            ((list? (atom-start (car res)))
             
             (let ((truth (eval (apply-sub sub
                                           (atom-start (car res)))
                                (interaction-environment))))
               (require+ truth)
               (if truth
                   (loop (cdr res) sub tag)
                   #f)))
            (else
             (let ((resolved (resolve res sub tag goal prog)))
               (if resolved
                   (loop (resolved->res resolved)
                         (resolved->sub resolved)
                         (+ tag 1))
                   #f)))))))





;;;
;;; WITH CUT
;;;

(define depth 0)
 
;; the symbol '! marks a cut.
(define cut?
  (lambda (x)
    (and (pair? x)
         (eqv? (car x) '!))))

(define prolog+cut
  (lambda (prog goal)
    
    (define asserta
      (lambda (res)
        (let ((fact (cdar res)))
          (set! prog (cons fact prog)))))
    
    (define retract
      (lambda (res)
        (let ((fact (cadar res)))
          (set! prog (remove (lambda (x) (equal? x fact))
                             prog)))))
 
    (amb+-reset)
    (amb+-add (lambda args #f))
    (let loop ((res (rename-vars goal #f))
               (sub '())
               (tag 1))
      
      (cond ((null? res)
             (select-sub sub goal))
           
            ;; check if it's a cut. If it is, set the AMB set of
            ;; continuations to res -- if we backtrack, we'll only
            ;; go this far back! 
            ((cut? (car res))
             (set! amb+-set (cadar res))
             (loop (cdr res) sub tag))
 
            ((eqv? (atom-start (car res)) 'asserta)
             (asserta res)
             (loop (cdr res) sub tag))
            
             ((eqv? (atom-start (car res)) 'retract)
             (retract res)
             (loop (cdr res) sub tag))
             
            ((eqv? (atom-start (car res)) 'is)
             (let ((s (do-is (car res) sub)))
               (loop (apply-sub s (cdr res))
                     (append s (apply-sub s sub)) tag)))
            
            ((list? (atom-start (car res)))
             (let ((r (eval
                       (apply-sub sub
                                  (atom-start
                                   (car res)))
                       (interaction-environment))))
               (require+ r)
               (if r
                   (loop (cdr res) sub tag)
                   #f)))
            (else
             (let ((resolved
                    (resolve res sub tag goal prog)))
               (if resolved
                   (loop (resolved->res resolved)
                         (resolved->sub resolved)
                         (+ tag 1))
                   #f)))))))

)


pll-standalone.scm
;; This software is Copyright (c) Jeronimo C. Pellegrini, 2010-2016.
;;
;;  This program is free software; you can redistribute it and/or
;;  modify it under the terms of the GNU General Public License as
;;  published by the Free Software Foundation; either version 3 of the
;;  License, or (at your option) any later version.
;; 
;;  This program is distributed in the hope that it will be useful, but
;;  WITHOUT ANY WARRANTY; without even the implied warranty of
;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;;  General Public License for more details.
;; 
;;  You should have received a copy of the GNU General Public License
;;  along with this program; if not, write to:
;;    The Free Software Foundation, Inc.,
;;    51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

;; pll-standalone.scm -- the loadable file for the Prolog interpreter,
;;                       as standalone program (not a R7RS library)

(cond-expand (mit         (define (interaction-environment) user-initial-environment))
	     (sisc        (require-extension (srfi 1)))
	     (chibi       (import (scheme r5rs))
			  (import (srfi 1))
			  (import (chibi show)))
	     (else))

(begin
  (load "amb.scm")
  (load "unify.scm")
  (load "prolog-core.scm"))



pll.scm
;; This software is Copyright (c) Jeronimo C. Pellegrini, 2010-2016.
;;
;;  This program is free software; you can redistribute it and/or
;;  modify it under the terms of the GNU General Public License as
;;  published by the Free Software Foundation; either version 3 of the
;;  License, or (at your option) any later version.
;; 
;;  This program is distributed in the hope that it will be useful, but
;;  WITHOUT ANY WARRANTY; without even the implied warranty of
;;  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
;;  General Public License for more details.
;; 
;;  You should have received a copy of the GNU General Public License
;;  along with this program; if not, write to:
;;    The Free Software Foundation, Inc.,
;;    51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

;; pll.scm -- the loadable file for the Prolog interpreter

(cond-expand (chicken  (require-extension r7rs))
             (else))

(define-library (pll)
  (import (scheme base) (scheme eval) (scheme repl) (scheme cxr) (scheme write) (srfi 1))
  ;; include-library-declarations didn't work as I was expecting on Chicken...
  (include  "amb.scm"
            "unify.scm"
            "prolog-core.scm")
  (export ;; AMB:
   fail amb-reset amb amb-save amb-restore require amb-all
   amb+-set amb+-reset amb+-fail amb+ amb+-list require+ amb-list
   ;; unification:
   matching-symbol? apply-sub variable? bound? occurs? unify
   ;; Prolog:
   pure-prolog
   prolog+built-ins
   prolog+scheme
   prolog+local
   prolog+meta
   prolog+cut))
